File: Graph/chaechon/7562.py
# 나이트의 이동
import sys
from collections import deque

input = sys.stdin.readline

def bfs(n,m):
    
    dx = [2,1,-1,-2,-2,-1,1,2]
    dy = [-1,-2,-2,-1,1,2,2,1]
    
    queue = deque()
    queue.append((n,m))
    
    visited[n][m] = 1
    
    while queue:
        
        a,b = queue.popleft()
        
        if a == dest_x and b == dest_y:
            return 0
        
        for i in range(8):
            nx = a + dx[i]
            ny = b + dy[i]
            
            if 0<=nx<size and 0<=ny<size:
                 
                if nx == dest_x and ny == dest_y:
                    visited[nx][ny] = 1
                    return chess[a][b] + 1
                
                elif visited[nx][ny] == 0:
                    chess[nx][ny] = chess[a][b] + 1
                    visited[nx][ny] = 1
                    queue.append((nx,ny))
                    # print(chess)
        
        
        
T = int(input())

for _ in range(T):
    size = int(input())
    start_x, start_y = map(int, input().split())
    dest_x, dest_y = map(int, input().split())
    
    chess = [[0]*size for _ in range(size)]
    visited = [[0]*size for _ in range(size)]
    # print(chess)
    print(bfs(start_x, start_y))
    
    

File: Graph/7562.kt
import java.util.LinkedList
import java.util.Queue

data class Pair(val x: Int, val y: Int, val count: Int)

fun main() = with(System.`in`.bufferedReader()) {
    val N = readLine().toInt()
    repeat(N) {
        val l = readLine().toInt()
        val (startX, startY) = readLine().split(" ").map { it.toInt() }
        val (endX, endY) = readLine().split(" ").map { it.toInt() }

        val visited = Array(l) { BooleanArray(l) }

        val dx = arrayOf(-1, -2, -2, -1, 1, 2, 2, 1)
        val dy = arrayOf(2, 1, -1, -2, -2, -1, 1, 2)

        fun bfs(startX: Int, startY: Int): Int {
            if (startX == endX && startY == endY) return 0

            val queue: Queue<Pair> = LinkedList()
            queue.add(Pair(startX, startY, 0))
            visited[startX][startY] = true

            while (queue.isNotEmpty()) {
                val (x, y, endCount) = queue.poll()!!
                if (x == endX && y == endY) {
                    return endCount
                }
                for (i in 0 until 8) {
                    val nx = x + dx[i]
                    val ny = y + dy[i]

                    if (nx in 0 until l && ny in 0 until l && !visited[nx][ny]) {
                        visited[nx][ny] = true
                        queue.add(Pair(nx, ny, endCount + 1))
                    }
                }
            }
            return -1
        }
        println(bfs(startX, startY))
    }
}
